File: python/tests/test_hash_map.py
import os
import sys
import unittest
from data_structures.hash_map import HashMap
sys.path.append(os.path.join(os.getcwd(), os.path.pardir))

class HashMapTest(unittest.TestCase):

    def test_add(self):
        print("Testing add")
        h = HashMap()

        h.add(1, "foo")
        h.add(2, "bar")
        h.add(3, "baz")

        self.assertEqual(h.get(1), "foo")
        self.assertEqual(h.get(2), "bar")
        self.assertEqual(h.get(3), "baz")
        self.assertEqual(h.get(0), None)

        #Chaining
        h.add(17, "foobar")
        self.assertEqual(h.get(17), "foobar")
        self.assertEqual(h.get(1), "foo")

    def test_remove(self):
        print("Testing remove")
        h = HashMap()

        h.add(1, "foo")
        h.add(2, "bar")
        h.add(3, "baz")

        self.assertEqual(h.get(1), "foo")
        h.remove(1)
        self.assertEqual(h.get(1), None)

        h.add(1, "foo")
        self.assertEqual(h.get(1), "foo")


        self.assertEqual(h.get(2), "bar")
        h.remove(2)
        self.assertEqual(h.get(2), None)

    def test_operator_get(self):
        print("Testing operator[] get")
        h = HashMap()

        h.add(1, "foo")
        h.add(2, "bar")
        h.add(3, "baz")

        self.assertEqual(h[1], "foo")
        self.assertEqual(h[2], "bar")
        self.assertEqual(h[4], None)
        self.assertEqual(h[0], None)

    def test_size(self):
        print("Testing size")
        h = HashMap()

        self.assertEqual(h.size, 0)
        h.add(1, "foo")
        self.assertEqual(h.size, 1)
        h.add(2, "bar")
        self.assertEqual(h.size, 2)
        h.remove(2)
        self.assertEqual(h.size, 1)
        h.remove(1)
        self.assertEqual(h.size, 0)

if __name__ == '__main__':
    unittest.main()

File: cpp/tests/test_hash_map.cpp
#include "catch.hpp"
#include "../data_structures/map/Bucket.h"
#include "../data_structures/map/hash_map/HashMap.cpp"
#include <string>

TEST_CASE("Test insert and []", "[HashSet]" ) {
    HashMap<int, int> intMap;
    HashMap<std::string, int> stringMap;

    for(int i = -1000; i < 1000; ++i) {
        intMap.insert(i, i+1);
    }
    for(int i = -1000; i < 1000; ++i) {
        REQUIRE(intMap.at(i) == i+1);
        REQUIRE(intMap[i] == i+1);
    }

    stringMap.insert("foo", 0);
    stringMap.insert("bar", 1);
    stringMap.insert("baz", 2);
    stringMap.insert("foobar", 3);
    stringMap.insert("foobarbaz", 4);

    REQUIRE(stringMap["foo"] == 0);
    REQUIRE(stringMap["bar"] == 1);
    REQUIRE(stringMap["baz"] == 2);
    REQUIRE(stringMap["foobar"] == 3);
    REQUIRE(stringMap["foobarbaz"] == 4);
}

TEST_CASE("Test contains and erase", "[HashSet]" ) {
    HashMap<int, int> intMap;
    HashMap<std::string, int> stringMap;

    for(int i = -1000; i < 1000; ++i) {
        intMap.insert(i, i+1);
        REQUIRE(intMap.contains(i));
    }
    for(int i = -1000; i < 1000; ++i) {
        intMap.erase(i);
        REQUIRE(!intMap.contains(i));
    }

    stringMap.insert("foo", 0);
    stringMap.insert("bar", 1);
    stringMap.insert("baz", 2);
    stringMap.insert("foobar", 3);
    stringMap.insert("foobarbaz", 4);

    REQUIRE(stringMap.contains("foo"));
    REQUIRE(stringMap.contains("bar"));
    REQUIRE(stringMap.contains("baz"));
    REQUIRE(stringMap.contains("foobar"));
    REQUIRE(stringMap.contains("foobarbaz"));

    stringMap.erase("bar");
    stringMap.erase("baz");
    stringMap.erase("foobar");

    REQUIRE(!stringMap.contains("bar"));
    REQUIRE(!stringMap.contains("baz"));
    REQUIRE(!stringMap.contains("foobar"));
}

TEST_CASE("Test size and empty for map", "[HashMap]" ) {
    HashMap<int, int> map;

    REQUIRE(map.empty());
    for(int i = 0; i < 1000; ++i) {
        REQUIRE(map.size() == i);
        map.insert(i, i);
        REQUIRE(!map.empty());
    }
    for(int i = 999; i >= 0; --i) {
        REQUIRE(!map.empty());
        map.erase(i);
        REQUIRE(map.size() == i);
    }
    REQUIRE(map.empty());
}
